src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

```diff
@@ -0,0 +1,275 @@
+    import TransformationError
+from psyclone.errors import InternalError
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE
+from psyclone.psyir.nodes import Range, Reference, ArrayReference, \
+    Assignment, Literal
+from psyclone.nemo import NemoLoop, NemoKern
+from psyclone.configuration import Config
+
+
+class NemoArrayRange2LoopTrans(Transformation):
+    '''Provides a transformation from a PSyIR ArrayReference Range to a
+    PSyIR NemoLoop. For example:
+
+    >>> schedule = ...
+    >>> range_node = ...
+    >>> from psyclone.domain.nemo.transformations import \
+    >>>     NemoArrayRange2LoopTrans
+    >>> trans = NemoArrayRange2LoopTrans()
+    >>> trans.apply(range_node)
+
+    '''
+    def apply(self, node, options=None):
+        '''Apply the NemoArrayRange2Loop transformation to the specified
+        node. The node must be a Range node within an array reference
+        on the lhs of an assignment. If the bounds of the Range are
+        specified then these are used. If they are not specified then
+        the lbound and ubound intrinsics are used unless the config
+        file has specified bounds, in which case these are used.
+
+        :param node: a Range node.
+        :type node: :py:class:`psyclone.psyir.nodes.Range`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dictionary of string:values or None
+
+        '''
+        self.validate(node)
+
+        array_reference = node.parent
+        array_index = array_reference.children.index(node)
+        assignment = array_reference.parent
+        parent = assignment.parent
+        symbol_table = node.scope.symbol_table
+
+        # See if there is any configuration information for this array index
+        loop_type_order = Config.get().api_conf("nemo").get_index_order()
+        loop_type_data = Config.get().api_conf("nemo").get_loop_type_data()
+        try:
+            loop_type = loop_type_order[array_index]
+            loop_type_info = loop_type_data[loop_type]
+            lower_bound_info = loop_type_info['start']
+            upper_bound_info = loop_type_info['stop']
+            loop_variable_name = loop_type_info['var']
+        except IndexError:
+            lower_bound_info = None
+            upper_bound_info = None
+            loop_variable_name = symbol_table.new_symbol_name("idx")
+
+        # Lower bound
+        if not array_reference.is_lower_bound(array_index):
+            # The range specifies a lower bound so use it
+            lower_bound = node.start
+        elif lower_bound_info:
+            # The config metadata specifies a lower bound so use it
+            try:
+                _ = int(lower_bound_info)
+                lower_bound = Literal(lower_bound_info, INTEGER_TYPE)
+            except ValueError:
```

Comment:
I think this may need a bit more care - e.g. what if a user has made a mistake editing the config file and got "1a" when they meant "1"? Does the configuration-file reader do any validation (i.e. check for valid integer or Fortran variable name)? If not, I think that's a new issue.

Reply:
No, the config-file reader just gets hold of the string and passes it on. I'll make a new issue and reference it.

Reply:
Issue is #1035